### PR TITLE
Use pure `solve` function

### PR DESCRIPTION
**Pull request by VictorCMiraldo**
Using a `pure` solve function and parallel symbolic evaluation seems promising; our state monad is hurting us a little, I might work on removing that state monad if not extremely necessary. 

Also, another curious fact is that keeping the lock in `pushMStack` and `popMStack` causes a deadlock, but removing it makes everything go through. I'm a bit puzzled by it.

### CODE REVIEW

**Comment by VictorCMiraldo** [comment]
@0xd34df00d, a quick google gave me this: https://stackoverflow.com/questions/23471594/concurrent-stack-implementation-using-mvar/23499300#23499300

**Comment by 0xd34df00d** [comment]
> @0xd34df00d, a quick google gave me this: https://stackoverflow.com/questions/23471594/concurrent-stack-implementation-using-mvar/23499300#23499300

Oh welp, noticed this just now, sorry! Anyway, using `QSem` seems cleaner than what's in that answer.

**Comment by serras** [comment]
Did GitHub get confused by the two PRs?

**Comment by VictorCMiraldo** [comment]
Either GitHub, Git or I got confused by the two PRs; I rebased this branch not inclusing the two initial commits on the latest `dev` (after merging #109) but now the web interface shows more commits than it should

**Comment by serras** [comment]
I think it's good now. I know that if you force-push GitHub leaves anything above it unchanged, but the Files tab work as it should.

File: src/PureSMT.hs [hunk 1]
```diff
@@ -22,3 +29,59 @@ class Solve domain where
...
+launchAll :: forall domain . Solve domain => Ctx domain -> IO [MVar X.Solver]
+launchAll ctx = replicateM numCapabilities $ do
+  s <- X.launchSolverWithFinalizer "cvc4 --lang=smt2 --incremental --fmf-fun" debug0
+  initSolver @domain ctx s
+  newMVar s
+  where
+    -- TODO: these constants should become parameters at some point; the solver command too!
+
+    numCapabilities :: Int
+    numCapabilities = 3
```
Comment (serras) [inline src/PureSMT.hs:63]: Why not start with the [RTS threading argument `numCapabilities`](https://hackage.haskell.org/package/base-4.14.1.0/docs/GHC-Conc.html#v:numCapabilities)?

File: src/Pirouette/Symbolic/Eval.hs [hunk 2]
```diff
@@ -327,19 +352,30 @@ symEvalOneStep t@(R.App hd args) = case hd of
           R.TermArg
           args
 
+-- A-ha... it's the state monad annoying us. Maybe if we can understand whether
+-- we really need the state monad we can better understand the parallelism possibilities
+-- in here. If we do really need the monad, we can only paralellize the prune and pruneAndValidate
+-- functions, if not, we could paralellize everything.
```
Comment (serras) [inline src/Pirouette/Symbolic/Eval.hs:358]: We are actually using `State` mostly to nest changes to the statistics. In fact, it should be possible to create a version of `ReaderT` which gives you back the "innermost" thing it used for its `local` or something like that.

**Review by serras — APPROVED** [review]

**Comment by VictorCMiraldo** [comment]
For the record; it's remarkable that Pirouette is significantly simpler and faster! This is CI's time for the `isUnity` test:
```
      isUnity Bug
        [correct_isUnity v] validate [\r _ -> r] counter:                                                 OK (2.71s)
```
A few PRs ago we were close to 10s! :tada:

**Pull request merged**